File: server/Agrigate.Core/Services/DeviceService/Models/DeviceDetails.cs
namespace Agrigate.Core.Services.DeviceService.Models;

/// <summary>
/// Detailed information about a specific device
/// </summary>
public class DeviceDetails
{
    public long Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string Location { get; set; } = string.Empty;
    public bool IsActive { get; set; }
    public List<RuleBase> Rules { get; set; } = [];
    public List<TelemetryBase> DistinctTelemetry { get; set; } = [];
}

File: agrigate/lib/models/devices/device_details.dart
import 'package:agrigate/models/devices/rule_base.dart';
import 'package:agrigate/models/devices/telemetry_base.dart';

class DeviceDetails {
  final int id;
  final String name;
  final String location;
  final bool isActive;
  final List<RuleBase> rules;
  final List<TelemetryBase> distinctTelemetry;

  const DeviceDetails({
    required this.id,
    required this.name,
    required this.location,
    required this.isActive,
    required this.rules,
    required this.distinctTelemetry,
  });

  factory DeviceDetails.fromJson(Map<String, dynamic> json) {
    return switch (json) {
      {
        'id': int id,
        'name': String name,
        'location': String location,
        'isActive': bool isActive,
        'rules': List<dynamic> rules,
        'distinctTelemetry': List<dynamic> telemetry,
      } =>
        DeviceDetails(
          id: id,
          name: name,
          location: location,
          isActive: isActive,
          rules: RuleBase.fromJsonList(rules),
          distinctTelemetry: TelemetryBase.fromJsonList(telemetry),
        ),
      _ => throw const FormatException('Failed to load DeviceDetails'),
    };
  }
}
